crates/matrix-sdk-crypto/src/session_manager/group_sessions.rs

```diff
@@ -847,6 +850,52 @@ impl GroupSessionManager {
 
         Ok(requests)
     }
+
+    /// Split the devices into two groups, one group for devices that we need to
+    /// share the room key with, and one group for devices that we don't
+    /// need to share the room key with.
+    fn split_devices_by_shared_with(
+        &self,
+        devices: BTreeMap<OwnedUserId, Vec<ReadOnlyDevice>>,
+        outbound: &OutboundGroupSession,
+    ) -> (Vec<ReadOnlyDevice>, Vec<(ReadOnlyDevice, ShareState)>) {
+        let split: (Vec<(ReadOnlyDevice, ShareState)>, Vec<(ReadOnlyDevice, ShareState)>) = devices
+            .into_iter()
+            .flat_map(|(_, d)| {
+                d.into_iter().map(|d| {
+                    let is_shared = outbound.is_shared_with(&d);
+                    (d, is_shared)
+                })
+            })
+            .partition(|(_, state)| matches!(state, ShareState::NotShared));
+
+        (split.0.into_iter().map(|(d, _)| d).collect(), split.1)
+    }
+
+    fn not_distributed_to_log_list(
+        &self,
+        already_shared_devices: &Vec<(ReadOnlyDevice, ShareState)>,
+        withthelds: &Vec<(ReadOnlyDevice, WithheldCode)>,
+    ) -> Vec<(String, String, String)> {
```

Comment:
A `Vec<(String, String, String)>` is quite the handful. Can we add a type with the desired `Debug` implementation instead? Then The `split_devices()` method can return that directly.